crates/starknet-devnet-core/src/state/state_readers.rs

```diff
@@ -0,0 +1,121 @@
+use std::collections::HashMap;
+
+use blockifier::execution::contract_class::ContractClass;
+use blockifier::state::cached_state::StorageEntry;
+use blockifier::state::errors::StateError;
+use blockifier::state::state_api::{State, StateReader, StateResult};
+use starknet_api::core::{ClassHash, CompiledClassHash, ContractAddress, Nonce};
+use starknet_api::hash::StarkFelt;
+use starknet_api::state::StorageKey;
+
+/// A simple implementation of `StateReader` using `HashMap`s as storage.
+/// Copied from blockifier test_utils, added `impl State`
+#[derive(Debug, Default, Clone)]
+pub struct DictState {
```

Comment:
is there a difference between DictState and DictStateReader? If no why not use it as an internal implementation of DictState. Like this: 
```rust
pub struct DictState(pub(crate) DictStateReader)
```

Reply:
The problem is in the ability to clone the struct. If I used `DictStateReader` as an internal struct, I'd still have to write my own `impl Clone`:
```rust
/// A simple implementation of `StateReader` using `HashMap`s as storage.
/// Copied from blockifier test_utils, added `impl State`
#[derive(Debug, Default)]
pub struct DictState(pub (crate) DictStateReader);

impl Clone for DictState {
    fn clone(&self) -> Self {
        Self(DictStateReader {
            storage_view: self.0.storage_view.clone(),
            address_to_nonce: self.0.address_to_nonce.clone(),
            address_to_class_hash: self.0.address_to_class_hash.clone(),
            class_hash_to_class: self.0.class_hash_to_class.clone(),
            class_hash_to_compiled_class_hash: self.0.class_hash_to_compiled_class_hash.clone(),
        })
    }
}
```

If we go this way, would we still have to provide the implementation of each `StateReader` method? By replacing e.g. `self.storage_view` with `self.0.storage_view`? Or would there be something smarter?

Reply:
By switching `pub struct DictState(pub(crate) DictStateReader);`, we would use less imports, but would need four more code lines. Not sure which way to go

Reply:
I see that both approaches have their tradeoffs. I will resolve this conversation